smithy-aws-cloudformation-traits/src/main/java/software/amazon/smithy/aws/cloudformation/traits/ResourceIndex.java

```diff
@@ -0,0 +1,479 @@
+    }
+
+    /**
+     * Get all members of the CloudFormation resource.
+     *
+     * @param resource ShapeID of a resource.
+     * @return Returns all members that map to CloudFormation resource
+     *   properties.
+     */
+    public Map<String, ResourcePropertyDefinition> getProperties(ToShapeId resource) {
+        return resourcePropertyMutabilities.getOrDefault(resource.toShapeId(), MapUtils.of())
+                .entrySet().stream()
+                .filter(entry -> !getExcludedProperties(resource).contains(entry.getValue().getShapeId()))
+                .collect(MapUtils.toUnmodifiableMap(Map.Entry::getKey, Map.Entry::getValue));
+    }
+
+    /**
+     * Gets the specified member of the CloudFormation resource.
+     *
+     * @param resource ShapeID of a resource
+     * @param propertyName Name of the property to retrieve
+     * @return The property definition.
+     */
+    public Optional<ResourcePropertyDefinition> getProperty(ToShapeId resource, String propertyName) {
+        return Optional.ofNullable(getProperties(resource).get(propertyName));
+    }
+
+    /**
+     * Get create-specifiable-only members of the CloudFormation resource.
+     *
+     * These properties can be specified only during resource creation and
+     * can be returned in a `read` or `list` request.
+     *
+     * @param resource ShapeID of a resource.
+     * @return Returns create-only member names that map to CloudFormation resource
+     *   properties.
+     */
+    public Set<String> getCreateOnlyProperties(ToShapeId resource) {
+        return getConstrainedProperties(resource, definition -> {
+            Set<Mutability> mutabilities = definition.getMutabilities();
+            return mutabilities.contains(Mutability.CREATE) && !mutabilities.contains(Mutability.WRITE);
+        });
+    }
+
+    /**
+     * Get read-only members of the CloudFormation resource.
+     *
+     * These properties can be returned by a `read` or `list` request,
+     * but cannot be set by the user.
+     *
+     * @param resource ShapeID of a resource.
+     * @return Returns read-only member names that map to CloudFormation resource
+     *   properties.
+     */
+    public Set<String> getReadOnlyProperties(ToShapeId resource) {
+        return getConstrainedProperties(resource, definition -> {
+            Set<Mutability> mutabilities = definition.getMutabilities();
+            return mutabilities.size() == 1 && mutabilities.contains(Mutability.READ);
+        });
+    }
+
+    /**
+     * Get write-only members of the CloudFormation resource.
+     *
+     * These properties can be specified by the user, but cannot be
+     * returned by a `read` or `list` request.
+     *
+     * @param resource ShapeID of a resource.
+     * @return Returns write-only member names that map to CloudFormation resource
+     *   properties.
+     */
+    public Set<String> getWriteOnlyProperties(ToShapeId resource) {
```

Comment:
I wonder if a lot of these methods can be moved to something like a `CfResource` type instead of kept on the index? That type would contain mutability, properties, exclusions, etc... Seems like that would greatly reduce the size of this index and give the definition more purpose IMO.

Reply:
This functionality was broken out to the `CfnResource` data class. Other related components were also moved to `Cfn` prefixed class names.